osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/TaikoDifficultyPreprocessor.cs

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Colour;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing
+{
+    public class TaikoDifficultyPreprocessor
+    {
+        /// <summary>
+        /// Creates a list of <see cref="TaikoDifficultyHitObject"/>s from a <see cref="IBeatmap"/>s.
+        /// This is placed here in a separate class to avoid <see cref="TaikoDifficultyCalculator"/> having to know
+        /// too much implementation details of the preprocessing, and avoid  <see cref="TaikoDifficultyHitObject"/>
+        /// having circular dependencies with various preprocessing and evaluator classes.
+        /// </summary>
+        /// <param name="beatmap">The beatmap from which the list of <see cref="TaikoDifficultyHitObject"/> is created.</param>
+        /// <param name="clockRate">The rate at which the gameplay clock is run at.</param>
+        public static List<DifficultyHitObject> CreateDifficultyHitObjects(IBeatmap beatmap, double clockRate)
```

Comment:
I don't really see the reason as to splitting this from the `TaikoDifficultyCalculator`, the added "implementation detail" of calling `TaikoColourDifficultyPreprocessor` in a line of code really doesn't seem like a big deal at alll.

Reply:
I'd be a little worried if this method got so complex to the point at which it'd need to be split off from the Difficulty Calculator, considering this is just for difficulty object creation. You've set precedence with calling the colour preprocessor from here to handle colour preprocessing, so I think it'd be fine to call some sort of general preprocessing class to handle all the complex work (maybe even this `TaikoDifficultyPreprocessor` class?)

I'm not sure how a circular dependency could come about from this method as well, unless I'm missing a hypothetical.

Reply:
Having `TaikoDifficultyCalculator` do the basic creation while having `TaikoDifficultyPreprocessor` handle further preprocessing makes sense actually. This was actually debated a bit when introduced due to `TaikoDifficultyCalculator.CreateDifficultyHitObjects` is essentially a proxy method for this now.

As for circular dependency it's not related to this preprocessor currently, but the preprocessor for colour, and potential future cases for other skills or even `TaikoDifficultyHitObject` (which `CreateDifficultyHitObject` was moved to previously). `TaikoColourDifficultyPreprocessor` prevents circular dependency between the `CoupledColourEncoding` and `TaikoDifficultyHitObjectColour`, due to `ProcessAndAssign` (which was previously in `CoupledColourEncoding`) needing access to `TaikoDifficultyHitObjectColour`. It also prevents circular dependencies due to pre-evaluation needing access to `ColourEvaluator` which in turns needs access to`TaikoDifficultyHitObjectColour` and the encoding classes. 

But I think it does make sense to have `TaikoDifficultyCalculator` to do the basic creation, and I will move it back there along with the implementation for non pre-evaluated colour difficulty.

Reply:
Applied in 7917a60e3c18c1f9f9d9c54fc7f1ae811b6cc470